### ISSUE TITLE
No documentation on how to upload photos

### ISSUE THREAD

**Issue by nicolas17** [mapillary/mapillary_apidoc#8]
There is no documentation on how to upload photos and sequences to Mapillary. There is only a vague reference saying `/me/uploads/secrets` can be used to get “user specific hashes to use S3 direct upload for images”, but there is no documentation on what to do with those hashes later. What S3 server do I upload to and how do I use these hashes? How do I assemble a sequence from the uploaded photos later?

The [Mapillary API v2](http://blog.mapillary.com/update/2015/08/05/apiv2.html) blog post says “The approach has been to design an API where the Mapillary webpage *or any of our apps* are simply users of the same API that is available to all developers.” Without photo upload documentation, it's not actually possible for developers to make a mobile app that collects images and submits them to Mapillary, like the official ones do.

**Comment by jesolem** [comment]
We have most of this on the blog but a bit spread out. Should of course be at the api doc. We will fix.

Thanks

**Comment by gyllen** [comment]
Hi @nicolas17 I have been waiting for this question. We are building an uploader at the moment and will document the processes as we move along.

**Comment by GITNE** [comment]
> Hi @nicolas17 I have been waiting for this question. We are building an uploader at the moment and will document the processes as we move along.

What kind of uploader? Like a library with bindings to different programming languages and runtimes? I would really like to know because I am currently improving the upload process of the [Mapillary plug-in for JOSM](https://github.com/floscher/josm-mapillary-plugin). It's kind of unstable, intransparent to the user, and does not recover from (network) errors. Ultimately, I would like to see the [Mapillary plug-in for JOSM](https://github.com/floscher/josm-mapillary-plugin) to become the one stop UI tool for professionals and advanced users (I am not saying that it is going to be me to do it, I would just like to see it happen). Juggling python scripts to pre-process and upload photos should not be the way to go in the long run.

**Comment by GITNE** [comment]
So, while attempting to improve the uploading experience in the Mapillary plug-in for JOSM, I started testing uploading against the Mapillary v2 API. Here is what I do and get from the server:
```
> POST /v2/me/uploads/images/CAFEBABE-0000-0000-0000-0000EBABEFAC/0000.jpg?client_id=T1Fzd20xZjdtR0s1VDk5OFNIOXpYdzoxNDYyOGRkYzUyYTFiMzgz HTTP/1.1
> User-Agent: Mapillary/32040 JOSM/1.5 (9973)
> Host: a.mapillary.com
> Accept: */*
> Authorization: Bearer <urlencoded oauth token>
> Content-Type: image/jpeg
> Content-Type-Encoding: 8bit
> Content-Length: 927325
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
<photo data/payload>
< HTTP/1.1 404 Not Found
< Server: openresty
< Date: Day, XX Mon XXXX XX:XX:XX GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 1351
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Access-Control-Request-Method: *
< Status: 404 Not Found
< X-Request-Id: XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
< X-Runtime: 0.790548
<
< <404 HTML page>
```
What do I do wrong? The API documentation mentions a :key (presumably a sequence key) in the POST URL. However, how do I get a sequence key for a new sequence? Or, how do I compute it? Please complete the API reference documentation.

**Comment by gyllen** [comment]
@GITNE Above is not how you are supposed to upload images. The full processed is not yet described in the API.

There is a reason for that,

We are at the moment building a native standalone uploader for Mapillary. During this process we might discover that we want to change the API. Once the uploader is built we will release the full upload process together with the source code of the uploader as a good example on how to build upon the upload API.

**Comment by GITNE** [comment]
>We are at the moment building a native standalone uploader for Mapillary. During this process we might discover that we want to change the API. Once the uploader is built we will release the full upload process together with the source code of the uploader as a good example on how to build upon the upload API.

Okay, fair enough. Unfortunately, this will stall my uploads to Mapillary and my development. Do you have a foreseeable time frame for when to expect the upload specification to become complete?

**Comment by GITNE** [comment]
Do you intend to release a Mapillary Java API (either closed or open source)? I mean, since you have an existing Android app, which of course may or may not be already written in Java, it my be fairly easy to do that. Otherwise, I am inclined to implement my own Java API, even though the Mapillary v2 API does not seem to be stable. Or, should I be rather patient? Any advice would be greatly appreciated.

**Comment by jesolem** [comment]
@GITNE We're considering options for what language libs to build. JavaScript and Python are high on the list, as is Java. 

If there are any parts of the v2 API you don't think are stable, let us know in an issue.

**Comment by gyllen** [comment]
@GITNE Im also intersted in what data in particular you are interested in from our API.

**Comment by GITNE** [comment]
> @GITNE We're considering options for what language libs to build. JavaScript and Python are high on the list, as is Java.

Sounds great! :+1: Thank you for the clarification.

> If there are any parts of the v2 API you don't think are stable, let us know in an issue.

What I meant was that based on the conversation the v2 API does not seem to be complete (or final) and is still subject to change, especially the upload part. The blog post seems to suggest that work on the v2 API has been completed. But, maybe it's just me and I got it wrong. :confused:

> @GITNE I'm also interested in what data in particular you are interested in from our API.

I am not sure I get the question right. As with regards to data, I am currently not interested in any particular or any additional data beyond what the API can or should deliver. However, I am interested in a working upload API call.

**Comment by moenk** [comment]
Me too. A simple example with curl would be fine.

**Comment by gyllen** [comment]
Have you guys had a look at https://github.com/mapillary/mapillary_tools, there are some examples in there, before I get around to do a wrtiteup

**Comment by GITNE** [comment]
> https://github.com/mapillary/mapillary_tools

Thank you for the hint. :smile: But, I do not think I want to incorporate external Python scripts into a Java plug-in, including a dependency on Python to run them. And, although I do not know the Python programming language, I could probably could reverse engineer the upload procedure from those scripts. But again, this is the wrong way to do software engineering and it would be a temporary and also error prone solution only. So, this is not a feasible way to go for me.

**Comment by moenk** [comment]
I think it is not a good practice to learn from the code, if there is an API it should be documented how to use it. 
As far as I understand I generate an UUID for each sequence and upload all pictures to a path with this UUID as a name, followed by a DONE file when complete. 
For an upload to S3 I need a bucket, a key and a secret. How can I get that? If we have that anybody can implement it according to existing S3 examples.

**Comment by nicolas17** [comment]
I agree with moenk. Reading the client source code only tells me what the client code does. Experimenting lets me guess what the server code does. API documentation says what part of the behavior of one side the other side can rely on. There is a big difference :)

Deriving protocol documentation by reading the code of one side doesn't always work well. An interesting example I saw once was open source client code that calculated a md5 hash, converted it to hexadecimal, and sent it to the server. By reading the client code alone, I wouldn't have guessed that the server only accepted *lowercase* hexadecimal. If there had been protocol documentation, either I would have known that clients must send it in lowercase, or I would have reported a server bug.

**Comment by gyllen** [comment]
Yes @nicolas17 and @moenk I do agree with you too. Let me see what I can do on documenting uploading.

**Comment by tobecwb** [comment]
Currently trying to upload several images to mapillary using python. Sure I can check the source code of mapillary_tools, but I agree with @nicolas17 and @moenk too! More than one year later, and I don't find any good documentation how I upload images.

**Comment by tobecwb** [comment]
@gyllen yes! I have the same opinion than @nicolas17  and @moenk. A good documentation it's better than a working code.
Today, I have to see all the code of the `upload_with_authentication.py` and the `uploader.py`, understand everything, to use less than 10% of all the example code.
And why Mapillary don't have a simple REST endpoint to send images? Why that mess with the `encode_multipart` function on `uploader.py` if request lib do almost everything automatically? so, my opinion it's the same! A good documentation it's better than a working code.

**Comment by nicolas17** [comment]
@gyllen The README has documentation on how to *use* the Python scripts. This GH issue is about having API documentation that would let me write a similar script from scratch. I don't *currently* have a need for that, but...

**Comment by gyllen** [comment]
@nicolas17 Yeah that makes total sense. We do not really expose that API since in general we want people to us the scripts to avoid errors. There has beens some talks lately though on that we should expose this API properly.

**Comment by GITNE** [comment]
> We do not really expose that API since in general we want people to us the scripts to avoid errors.

@gyllen By that logic you should not expose any API at all because every time somebody uses the API they can make mistakes or cause errors. :confused: Or, is it better for people to rely on a crude reimplementation of the script behavior than on a clean specification of the protocol? Besides, you do not prevent errors from happening just by saying “People, we want you to use X” (and hope for the best) in an open data environment but by a clean protocol specification, data integrity and validity checks on the server side.

> Why that mess with the `encode_multipart` function on `uploader.py` if request lib do almost everything automatically?

I can basically agree here because the multipart scheme is way too complex and too inefficient for what it is supposed to do. Uploading *many* files should be simpler.

**Comment by tobecwb** [comment]
For anybody that want to know how to upload a photo or a sequence, check this link. It's not a official documentation, but it's the best I can do. Included a CURL example.

https://gist.github.com/tobecwb/9c43b88f4caa2484c8086d8d01648b23
